common/cuda_hip/matrix/csr_kernels.hpp.inc

```diff
@@ -1046,3 +1046,70 @@ __global__ __launch_bounds__(default_block_size) void inv_symm_permute_kernel(
         out_vals[out_begin + i] = in_vals[in_begin + i];
     }
 }
+
+
+namespace kernel {
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void get_row_nnz_data(
+    const size_type num_rows, const size_type* __restrict__ nnz_per_row,
+    IndexType* __restrict__ row_ptrs)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    if (tidx < num_rows) {
+        row_ptrs[tidx] = nnz_per_row[tidx];
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void compute_submatrix_idxs_and_vals(
+        const size_type num_rows, const size_type num_cols,
+        const size_type num_nnz, const size_type row_offset,
+        const size_type col_offset,
+        const IndexType* __restrict__ source_row_ptrs,
+        const IndexType* __restrict__ source_col_idxs,
+        const ValueType* __restrict__ source_values,
+        const IndexType* __restrict__ result_row_ptrs,
+        IndexType* __restrict__ result_col_idxs,
+        ValueType* __restrict__ result_values)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    if (tidx < num_rows) {
+        size_type res_nnz = result_row_ptrs[tidx];
+        for (size_type nnz = source_row_ptrs[row_offset + tidx];
+             nnz < source_row_ptrs[row_offset + tidx + 1]; ++nnz) {
+            if ((source_col_idxs[nnz] < (col_offset + num_cols) &&
+                 source_col_idxs[nnz] >= col_offset)) {
+                result_col_idxs[res_nnz] = source_col_idxs[nnz] - col_offset;
+                result_values[res_nnz] = source_values[nnz];
+                res_nnz++;
+            }
+        }
+    }
+}
+
+
+template <typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void calculate_nnz_per_row_in_span(
+        const span row_span, const span col_span,
+        const IndexType* __restrict__ row_ptrs,
+        const IndexType* __restrict__ col_idxs,
+        size_type* __restrict__ nnz_per_row)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    if (tidx >= row_span.begin && tidx < row_span.end) {
```

Comment:
you could avoid some idle threads here for small submatrix extraction (need to adjust launch parameters, of course)
```suggestion
    const auto tidx = thread::get_thread_id_flat() + row_span.begin;
    if (tidx < row_span.end) {
```